go/private/context.bzl

```diff
@@ -785,7 +785,12 @@ cgo_context_data = rule(
             default = "@bazel_tools//tools/osx:current_xcode_config",
         ),
     },
-    toolchains = ["@bazel_tools//tools/cpp:toolchain_type"],
+    toolchains = [
+        # In pure mode, a C++ toolchain isn't needed when transitioning.
+        # But if we declare a mandatory toolchain dependency here, a cross-compiling C++ toolchain is required at toolchain resolution time.
+        # So we make this toolchain dependency optional, so that it's only attempted to be looked up if it's actually needed.
```

Comment:
Could you give an example of the error users would see if they need one but it isn't available? I want to make sure that the rest of the code handles the absence of the formerly required toolchain gracefully.

Reply:
Interesting, I went through a large matrix of:
Bazel version: 5.3.2 vs 6.0.0rc1 (i.e. doesn't have vs does have support for optional toolchains)
`--incompatible_enable_cc_toolchain_resolution` on or off
Target's `pure` on or off
Actually having a cross-compiling C++ toolchain registered or not

I saw the exact same behaviour without and with this PR except for the following cells:

* (The intended change): Bazel 6.0.0rc1 with no cross-compiling C++ toolchain with `--incompatible_enable_cc_toolchain_resolution` off building a `pure="on"` target: Instead of a toolchain error (behaviour on master), we get a successful build
* (Surprising): Bazel 6.0.0rc1 with no cross-compiling C++ toolchain with `--incompatible_enable_cc_toolchain_resolution` off building a `pure="off"` target: Instead of a toolchain error (behaviour on master), we get a build-time error because a C++ toolchain _is_ found, but it's not a cross-compiling one, so an invalid command line is run.

I suspect this change probably isn't worth merging in that case - particularly as I'm a believer in the `--incompatible_enable_cc_toolchain_resolution` future, and this only changes behaviour when that flag is disabled.

Reply:
The surprising case is indeed... surprising. Could you add details about the toolchain that is found in that case? If rules_go is behaving incorrectly in this case, we should fix it. If Bazel does, that would make optional toolchains quite a bit less useful.

Reply:
Absolutely - I just pushed a new commit

If on a mac (I'm on an x86_64 machine), you run `USE_BAZEL_VERSION=6.0.0rc1 bazel build //tests/core/cross:linux_go_cross_impure_cgo` you should see Bazel hand rules_go a darwin->darwin C++ toolchain, which then fails to build the stdlib for linux like this:

```
ERROR: /Volumes/Source/github.com/bazelbuild/rules_go/BUILD.bazel:42:7: GoStdlib stdlib_/pkg failed: (Exit 1): builder failed: error executing command (from target //:stdlib) bazel-out/darwin-opt-exec-2B5CBBC6-ST-759b62b2e9d9/bin/external/go_sdk/builder stdlib -sdk external/go_sdk -installsuffix linux_amd64 -out bazel-out/darwin-fastbuild-ST-79e94b232e60/bin/stdlib_

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
# runtime/cgo
linux_syscall.c:67:13: error: implicit declaration of function 'setresgid' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
linux_syscall.c:67:13: note: did you mean 'setregid'?
[redacted]/usr/include/unistd.h:593:6: note: 'setregid' declared here
linux_syscall.c:73:13: error: implicit declaration of function 'setresuid' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
linux_syscall.c:73:13: note: did you mean 'setreuid'?
[redacted]/usr/include/unistd.h:595:6: note: 'setreuid' declared here
Error in child process '/usr/bin/xcrun'. 1
stdlib: error running subcommand external/go_sdk/bin/go: exit status 2
Target //tests/core/cross:linux_go_cross_impure_cgo failed to build
Use --verbose_failures to see the command lines of failed build steps.
```

Reply:
Okay, that actually isn't that surprising: Without toolchain resolution, the transition in `go_cross_binary` is a no-op and it makes sense that the auto-configured host toolchain is used.

Do you know why you didn't see a difference with `--incompatible_enable_cc_toolchain_resolution` *enabled*? I agree that is the interesting case.

Reply:
Tried this myself, I think I have an idea of what is wrong:

1. https://github.com/bazelbuild/rules_go/blob/1a8fe64877c6e71dbbf51cdc8ceb2eb10c13e521/go/private/context.bzl#L779 uses the legacy way to get the C++ toolchain, which always fails if none is available even if it isn't marked as mandatory. 
2. Removing this line changes the error to one in the Starlark `find_cpp_toolchain` helper, which also always fails if the toolchain isn't available. 

I will try what happens if I use a helper that is aware of optional toolchains.

Reply:
@katre Do you see a way for rules to both work without C++ toolchain resolution (using `current_cc_toolchain`) *and* make use of optional toolchains? It seems as if depending on `@bazel_tools//tools/cpp:current_cc_toolchain` is necessary for the former, but breaks the latter.

Reply:
@illicitonion Using `cc_toolchain = ctx.toolchains["@bazel_tools//tools/cpp:toolchain_type"]` directly in `_cgo_context_data_impl` correctly returns `None` in the cross-compilation case with `--incompatible_enable_cc_toolchain_resolution`. Not sure though we can get this behavior without losing support for non-toolchain resolution builds, given the above.

Reply:
@fmeum You may need to import/copy the latest version of [`find_cc_toolchain.bzl`](https://github.com/bazelbuild/rules_cc/blob/main/cc/find_cc_toolchain.bzl), which correctly handles all these options.

Reply:
I did that, but with the commands provided by @illicitonion and `--incompatible_enable_cc_toolchain_resolution`, I get a Java error in the native `cc_toolchain_alias` rule even before that Starlark function is executed. It only goes away if I remove the `_cc_toolchain` attribute with default `@bazel_tools//tools/cpp:current_cc_toolchain`, but that would break backwards compatibility.